Volume and area preserving barostats option

Bibobu opened the pull request:
**Summary**

This is an implementation of an option aimed at solving #4623 . I tried to take the discussion from #4419 into account.
The PR adds a keyword `isochoric` to NPT simulation that splits the strain from barostated dimensions on non-barostated dimensions. For now, the split is equal on remaining dimensions. The choice of `isochoric` keyword should avoid the confusion with `volume` or `couple` keywords by using a specific technical term from thermodynamics.

**Related Issue(s)**

As mentioned #4623 and #4419.

**Author(s)**

Germain Clavier (Université Caen Normandie, germain.clavier@unicaen.fr)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

This PR should not break backward compatibility.

**Implementation Notes**

The code was tested on an FCC argon crystal at 30K and 1 atm.

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

Bibobu committed:
Merge branch 'develop' of github.com:lammps/lammps into develop

Bibobu committed:
Merge branch 'develop' of github.com:lammps/lammps into develop

Bibobu committed:
Merge branch 'develop' of github.com:lammps/lammps into develop

Bibobu committed:
Merge branch 'fix-gauss' of github.com:martok/lammps into develop

Bibobu committed:
Merge branch 'develop' of github.com:lammps/lammps into develop

Bibobu committed:
Merge branch 'develop' of github.com:lammps/lammps into develop

Bibobu committed:
Merge branch 'develop' of github.com:lammps/lammps into develop

Bibobu committed:
Added some variable to fix_nh

Bibobu committed:
Added the isochoric flag to fix_nh.h

Bibobu committed:
Works better with ;

Bibobu committed:
Added isochoric option in fix_nh.cpp

Bibobu committed:
Added documentation for fix_nh isochoric option

Bibobu committed:
Added some details in isochoric keyword documentation

Bibobu committed:
Merge branch 'develop' of github.com:lammps/lammps into NPTisochoric

Bibobu committed:
Added isochoric default value to doc.

Bibobu committed:
Merge branch 'develop' of github.com:lammps/lammps into NPTisochoric

stanmoore1 commented:
@Bibobu any update on this PR?

Bibobu commented:
> @Bibobu any update on this PR?

This actually waits for reviews since the project was left off by @srtee. Is there anything more required?

akohlmey commented:
> Is there anything more required?

@Bibobu at the very minimum, the branch needs to pass all tests and ideally also should be up-to-date with upstream.

athomps reviewed (commented):
@Bibobu This is a nice feature.  I hope you can bring it up to date and resolve test errors.  The `fix box/deform` command has a `volume` style that can be specified separately for x, y, and z dimensions and triggers the same functionality, but with more selective power e.g. if I am barostatting a polymer in x, and ly is very small but lz is large, I may want to only adjust lz to preserver volume, while keeping ly fixed ..... `x 1.0 1.0 0.5....z volume`

Bibobu commented:
@akohlmey I didn't expect any programming style error. I'll check the code again. However, I don't know where the windows failure comes from.

@athomps do you think this would require a similar feature? Not a general flag but a `volume` flag per-dimension? In that case `isochoric` keyword could be a shortcut just like `iso` `aniso` or `tri`.

akohlmey committed:
Merge branch 'develop' into NPTisochoric

akohlmey commented:
> @akohlmey I didn't expect any programming style error. I'll check the code again.

Please see: https://docs.lammps.org/Build_development.html#coding-style-utilities

akohlmey commented:
> However, I don't know where the windows failure comes from.

I have updated the baseline from upstream and will check it out, if there are still issues.

akohlmey commented on src/fix_nh.cpp:
```diff
@@ -483,6 +505,13 @@ FixNH::FixNH(LAMMPS *lmp, int narg, char **arg) :
+    if (isochoric) {
+      if (not p_flag[0]) box_change |= BOX_CHANGE_X;
+      if (not p_flag[1]) box_change |= BOX_CHANGE_Y;
+      if (dimension == 3 && not p_flag[2]) box_change |= BOX_CHANGE_Z;
```
```suggestion
      if (dimension == 3 && !p_flag[2]) box_change |= BOX_CHANGE_Z;
```

@Bibobu here is the Windows issue.
Using "not" instead of "!" is not portable to MS Visual C++.

srtee commented:
> > @Bibobu any update on this PR?
> 
> This actually waits for reviews since the project was left off by @srtee. Is there anything more required?

For context, I did the initial work very quickly in response to a MatSci post and some random extra time for thought. I'm not using this feature in a continuing way so I won't be able to comment on how this feature is progressing (and I am happy for the credit to go solely to @Bibobu if that suits the LAMMPS contrib style better, and definitely mainly to @Bibobu).

akohlmey committed:
Merge branch 'develop' into NPTisochoric

akohlmey committed:
replace non-portable text logical operator with traditional one

akohlmey commented:
@Bibobu I've fixed portability and whitespace issues and updated it to the latest upstream.

Please let us know if this has any chance to be completed soon.
Otherwise, I suggest to close the pull request and revisit it at some later point.
For as long as there is a reference to this pull request in the issue, anybody else who wants to take a closer look can start from this change set instead of starting from scratch.

srtee commented:
Hi @akohlmey and @Bibobu I'll look at this and finalise by the end of this week (i.e. 168 hours from this message), otherwise let's close it for now.

Cheers, Shern

Bibobu commented:
I should also have some time to work on it this week. Same answer as @srtee

Bibobu committed:
Merge branch 'NPTisochoric' of github.com:Bibobu/lammps into NPTisochoric

Bibobu committed:
Added p_isoch array for isochoric option

Bibobu committed:
Refactored isochoric option to allow for dimension choice

Bibobu committed:
Rephrased the documentation of isochoric option according to changes.

Bibobu committed:
Typo

Bibobu committed:
Corrected periodic boundary check for isochoric dimensions

Bibobu commented:
These last changes should take @athomps remarks into account. It simplifies the logic a lot by introducing a flag for each isochoric dimension. As a maximum of 2 can be used at the same time, there is a limited amount of checks to be done with regards to other NPT parameters. It is now possible to choose which dimensions to use to compensate for the strain induced by the barostat. It is also possible to have non-periodic dimensions which was not possible in the previous version of the implementation.

The documentation was rephrased to better reflect this behavior. I shall try to write some small tests soon.

jtclemm commented on doc/src/fix_nh.rst:
```diff
@@ -375,6 +376,24 @@ acquire ghost atoms around a processor's irregular-shaped subdomain.
+dimensions to use in that regard: "x" indicates the x dimension, "yz" (no
+space) indicates the y and z dimensions. The selected dimensions are scaled to
+compensate the strain of the barostat and keep the system at a constant volume
+(or surface in 2d). It is not possible to use this keyword if all the
```
Maybe area instead of surface is clearer?

jtclemm commented on doc/src/fix_nh.rst:
```diff
@@ -375,6 +376,24 @@ acquire ghost atoms around a processor's irregular-shaped subdomain.
 error.
 
+The *isochoric* keyword allows to maintain constant volume during NPT
+simulations. The values following the isochoric dimension indicates the
```
allows NPT simulations at constant volume?

Bibobu replied:
Ah yes, this phrasing doesn't make much sense as is. I'll rewrite this sentence asap.

jtclemm commented on src/fix_nh.cpp:
```diff
@@ -1170,12 +1223,41 @@ void FixNH::remap()
+      if (scalexz) h[4] *= isofac;
+      if (scaleyz) h[3] *= isofac;
+    }
+  }
```
Over long times, wouldn't this protocol create a drift in the volume due to accumulating machine precision errors? Would it make sense to alternatively save the original volume and use that to rescale isochoric dimensions?

Bibobu replied:
Good remark. My initial idea was simply to split instantaneous strain on the remaining directions and implementing the direction selection already simplified the algorithm. I didn't think about it that way.

I should give it a thought. I can't tell if this would become an issue for long runs at the moment.

jtclemm replied:
In case it's useful, here's the corresponding *volume* command from fix deform that (from my understanding) operates this way:
https://github.com/lammps/lammps/blob/0609e9b1665941b681f05963a08ab4cefb3c0d23/src/fix_deform.cpp#L1055

`set.vol_start` is set in `init()` and the `set` structures are written/read from restart values so users can continue constant volume deformations.

Bibobu committed:
Rephrased *isochoric* keyword explanation.

Clarified the usage of the *isochoric* keyword in NPT simulations by making the description of the scaling process clearer.

Bibobu committed:
Changed the algorithm to recompute scaling factor with regard to reference volume.

Bibobu committed:
Added isochoric NPT option to the OPENMP nosé-hoover version.

Bibobu committed:
Merge branch 'NPTisochoric' of github.com:Bibobu/lammps into NPTisochoric

Bibobu commented:
The last PR took the comment of @jtclemm into account and now recompute the scaling factors with regard to the reference volume which is the volume at the moment the fix is initialized. I also made the change in the OPENMP version of nh and put the data in the restart related methods. However I still have issues with the tests set that I couldn't trace down yet (didn't have time last week). This was only related to the restart files for now and maybe my test is badly written. I'll try to investigate in the coming week but I'll push as is since it works in the general case.

Bibobu committed:
Merge branch 'develop' into NPTisochoric

jtclemm commented:
Nice! The changes in this PR look good to me once you have the chance to track down your failing test.

akohlmey committed:
Merge branch 'develop' into NPTisochoric

akohlmey commented:
> However I still have issues with the tests set that I couldn't trace down yet (didn't have time last week).

Is this using YAML files not included in the branch?
The automatic tests with the included files seem to be passing.
If you include any additional YAML files into the branch, I can have a look, too.

Bibobu committed:
Added falsy test file for isochoric npt option

Bibobu committed:
Merge branch 'NPTisochoric' of github.com:Bibobu/lammps into NPTisochoric

Bibobu commented:
> Is this using YAML files not included in the branch?

It did. I've included the file I've written in unittest in the last commit so you can have a look. When running the tests it is still failing with the following output:

```

Output:
----------------------------------------------------------
[==========] Running 3 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 3 tests from FixTimestep
[ RUN      ] FixTimestep.plain
/home/germain/Documents/Codes/lammps/unittest/force-styles/test_main.cpp:98: Failure
Expected: (err) <= (epsilon)
  Actual: 2.5351601210898602e-12 vs 2e-12
Google Test trace:
/home/germain/Documents/Codes/lammps/unittest/force-styles/test_main.cpp:89: EXPECT_VELOCITIES: run_vel (normal run, verlet)

/home/germain/Documents/Codes/lammps/unittest/force-styles/test_fix_timestep.cpp:329: Failure
Expected: (err) <= (epsilon)
  Actual: 6 vs 2e-12

/home/germain/Documents/Codes/lammps/unittest/force-styles/test_main.cpp:98: Failure
Expected: (err) <= (epsilon)
  Actual: 2.0815885095540655e-12 vs 2e-12
Google Test trace:
/home/germain/Documents/Codes/lammps/unittest/force-styles/test_main.cpp:89: EXPECT_VELOCITIES: run_vel (restart, verlet)

/home/germain/Documents/Codes/lammps/unittest/force-styles/test_fix_timestep.cpp:381: Failure
Expected: (err) <= (epsilon)
  Actual: 4 vs 2e-12

/home/germain/Documents/Codes/lammps/unittest/force-styles/test_main.cpp:98: Failure
Expected: (err) <= (epsilon)
  Actual: 2.0815885095540655e-12 vs 2e-12
Google Test trace:
/home/germain/Documents/Codes/lammps/unittest/force-styles/test_main.cpp:89: EXPECT_VELOCITIES: run_vel (rmass, verlet)

/home/germain/Documents/Codes/lammps/unittest/force-styles/test_fix_timestep.cpp:419: Failure
Expected: (err) <= (epsilon)
  Actual: 4 vs 2e-12

/home/germain/Documents/Codes/lammps/unittest/force-styles/test_fix_timestep.cpp:477: Failure
Expected: (err) <= (10 * epsilon)
  Actual: 6 vs 2.0000000000000001e-09

/home/germain/Documents/Codes/lammps/unittest/force-styles/test_fix_timestep.cpp:514: Failure
Expected: (err) <= (10 * epsilon)
  Actual: 4 vs 2.0000000000000001e-09

/home/germain/Documents/Codes/lammps/unittest/force-styles/test_fix_timestep.cpp:552: Failure
Expected: (err) <= (10 * epsilon)
  Actual: 4 vs 2.0000000000000001e-09

[  FAILED  ] FixTimestep.plain (36 ms)
[ RUN      ] FixTimestep.omp
```

I think the output is also different compared to the previous implementation without initial volume stored. But I didn't have time yet to investigate further.

akohlmey commented:
> Actual: 2.5351601210898602e-12 vs 2e-12

This error is due to precision and can be avoided by changing the epsilon parameter to a slightly larger value. This is normal when adding some extra processing to data. The tests are designed to be very sensitive by default, but can be "blunted" by adjusting the epsilon value. For most cases, epsilon can be close to the floating point epsilon, but often they need to be adjusted by a couple orders of magnitude (depending on the number of floating point operations that go into computing the property of interest). There is, however, a limit where the tests become too insensitive and where it can be difficult to tell the difference between a problem and accepted numerical noise.

> /home/germain/Documents/Codes/lammps/unittest/force-styles/test_fix_timestep.cpp:419: Failure
> Expected: (err) <= (epsilon)
>   Actual: 4 vs 2e-12
> 
> /home/germain/Documents/Codes/lammps/unittest/force-styles/test_fix_timestep.cpp:477: Failure
> Expected: (err) <= (10 * epsilon)
>   Actual: 6 vs 2.0000000000000001e-09

These are a problem that needs looking into.

akohlmey committed:
flush small values for stress or global scalars or vectors to zero to avoid false positives

akohlmey committed:
update test reference

akohlmey commented:
@Bibobu I've addressed the issues with the plain version of the tests. It was due to a deficiency in the tester program. It should force very small values for stress or global scalars/vectors to zero to avoid false positives. With very small values any *relative* errors become huge. With the changed epsilon, the tests pass for the plain version. Both KOKKOS/OpenMP and OPENMP are still failing. I have - temporarily - disabled the tests in the YAML file (see `skip_tests:`). This looks more like a general problem of something missing. You can either disable the feature for KOKKOS and OPENMP and leave it for later, or re-check your port and then re-enable the tests.

athomps reviewed (approved):
I approve.
